updateDynamic with implicit conversion from String for "name" param

Given a Dynamic class/trait, with updateDynamic(name: Something)(...), and an implicit conversion from String to Something, the typechecker will report "reassignment to val" when trying to update a field using the = syntax. It does work fine for selectDynamic and applyDynamic.

Minimized test case:

```scala
import scala.language.{ dynamics, implicitConversions }

class ID(s: String)

class DynamicFields extends scala.Dynamic {
  def updateDynamic(name: ID)(value: Any): Unit = ???
}

object BugUpdateDynamic {
  implicit def string2id(s: String): ID = new ID(s)

  def bug(): Unit = {
    val o = new DynamicFields
    o.foo = "bar"
  }
}
```

Compiling with 2.10.1, 2.10.2 or 2.11.0-M3 yields the same output.

Expected output: compile fine, without error.

Actual output:

```scala
$ ~/opt/scala-2.11.0-M3/bin/scalac BugUpdateDynamic.scala
BugUpdateDynamic.scala:14: error: reassignment to val
    o.foo = "bar"
          ^
one error found
```

(manually calling updateDynamic works, btw)

--- a/src/compiler/scala/tools/nsc/typechecker/Typers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
@@ -4169,10 +4169,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       if (!isApplyDynamicName(name) && acceptsApplyDynamic(qual.tpe.widen)) Some(NoType)
       else None
 
+    // if the qualifier is a Dynamic, that's all we need to know
     private def isDynamicallyUpdatable(tree: Tree) = tree match {
-      // if the qualifier is a Dynamic, that's all we need to know
       case DynamicUpdate(qual, name) => acceptsApplyDynamic(qual.tpe)
-      case _ => false
+      case _                         => false
     }
 
     private def isApplyDynamicNamed(fun: Tree): Boolean = fun match {
@@ -4445,7 +4445,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           val rhs1 = typedByValueExpr(rhs, lhs1.tpe)
           treeCopy.Assign(tree, lhs1, checkDead(context, rhs1)) setType UnitTpe
         }
-        else if(isDynamicallyUpdatable(lhs1)) {
+        else if (isDynamicallyUpdatable(lhs1)) {
           val t = atPos(lhs1.pos.withEnd(rhs.pos.end)) {
             Apply(lhs1, List(rhs))
           }

--- a/src/compiler/scala/tools/nsc/typechecker/Typers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
@@ -959,11 +959,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
 
       def insertApply(): Tree = {
         assert(!context.inTypeConstructorAllowed, mode) //@M
-        val adapted = adaptToName(tree, nme.apply)
+        val adapted = adaptToName(unmarkDynamicRewrite(tree), nme.apply)
         val qual = gen.stabilize(adapted)
-        typedPos(tree.pos, mode, pt) {
-          Select(qual setPos tree.pos.makeTransparent, nme.apply)
-        }
+        val t = atPos(tree.pos)(Select(qual setPos tree.pos.makeTransparent, nme.apply))
+        wrapErrors(t, _.typed(t, mode, pt))
       }
       def adaptConstant(value: Constant): Tree = {
         val sym = tree.symbol
@@ -1114,12 +1113,19 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           def hasPolymorphicApply = applyMeth.alternatives exists (_.tpe.typeParams.nonEmpty)
           def hasMonomorphicApply = applyMeth.alternatives exists (_.tpe.paramSectionCount > 0)
 
-          acceptsApplyDynamic(tree.tpe) || (
-            if (mode.inTappMode)
-              tree.tpe.typeParams.isEmpty && hasPolymorphicApply
-            else
-              hasMonomorphicApply
-          )
+          def badDynamicApply() = {
+            tree match {
+              case Apply(fun, _) => DynamicRewriteError(tree, ApplyWithoutArgsError(tree, fun))
+              case _             => ()
+            }
+            false
+          }
+          if (acceptsApplyDynamic(tree.tpe))
+            !isDynamicRewrite(tree) || badDynamicApply()
+          else if (mode.inTappMode)
+            tree.tpe.typeParams.isEmpty && hasPolymorphicApply
+          else
+            hasMonomorphicApply
         }
         def shouldInsertApply(tree: Tree) = mode.typingExprFun && {
           tree.tpe match {

--- a/src/reflect/scala/reflect/internal/TreeInfo.scala
+++ b/src/reflect/scala/reflect/internal/TreeInfo.scala
@@ -903,12 +903,19 @@ abstract class TreeInfo {
 
   def isApplyDynamicName(name: Name) = (name == nme.updateDynamic) || (name == nme.selectDynamic) || (name == nme.applyDynamic) || (name == nme.applyDynamicNamed)
 
+  private object LiteralNameOrAdapted {
+    def unapply(tree: Tree) = tree match {
+      case Literal(Constant(name))                 => Some(name)
+      case Apply(_, List(Literal(Constant(name)))) => Some(name)
+      case _                                       => None
+    }
+  }
   class DynamicApplicationExtractor(nameTest: Name => Boolean) {
     def unapply(tree: Tree) = tree match {
-      case Apply(TypeApply(Select(qual, oper), _), List(Literal(Constant(name)))) if nameTest(oper) => Some((qual, name))
-      case Apply(Select(qual, oper), List(Literal(Constant(name)))) if nameTest(oper) => Some((qual, name))
-      case Apply(Ident(oper), List(Literal(Constant(name)))) if nameTest(oper) => Some((EmptyTree, name))
-      case _ => None
+      case Apply(TypeApply(Select(qual, oper), _), List(LiteralNameOrAdapted(name))) if nameTest(oper) => Some((qual, name))
+      case Apply(Select(qual, oper), List(LiteralNameOrAdapted(name))) if nameTest(oper)               => Some((qual, name))
+      case Apply(Ident(oper), List(LiteralNameOrAdapted(name))) if nameTest(oper)                      => Some((EmptyTree, name))
+      case _                                                                                           => None
     }
   }
   object DynamicUpdate extends DynamicApplicationExtractor(_ == nme.updateDynamic)

--- a/test/files/neg/t9310.check
+++ b/test/files/neg/t9310.check
@@ -1,0 +1,6 @@
+t9310.scala:8: error: (arg: Any)Test.this.C does not take parameters
+error after rewriting to Test.this.c.applyDynamic("m")
+possible cause: maybe a wrong Dynamic method signature?
+  def f = c.m(42)
+          ^
+one error found

--- a/test/files/neg/t9310.scala
+++ b/test/files/neg/t9310.scala
@@ -1,0 +1,28 @@
+import scala.language.dynamics
+
+trait Test {
+  class C extends Dynamic {
+    def applyDynamic(arg: Any): C = ???
+  }
+  val c = new C
+  def f = c.m(42)
+}
+
+/*
+object Test {
+  case class LeonAny(v: Any) extends Dynamic {
+    def applyDynamic(args: Any*): LeonAny = this
+  }
+
+  implicit def fooToFoo(a: Any): LeonAny = {
+    LeonAny(a)
+  }
+
+  def test() {
+    val a: LeonAny = 42
+    val b: LeonAny = Nil
+
+    b += a
+  }
+}
+*/

--- a/test/files/pos/t7606.scala
+++ b/test/files/pos/t7606.scala
@@ -1,0 +1,20 @@
+import scala.language.{ dynamics, implicitConversions }
+
+class ID(s: String)
+
+class DynamicFields extends scala.Dynamic {
+  def updateDynamic(name: ID)(value: Any): Unit = ???
+}
+
+object BugUpdateDynamic {
+  implicit def string2id(s: String): ID = new ID(s)
+
+  def explicitly(): Unit = {
+    val o = new DynamicFields
+    o.updateDynamic("foo")("bar")
+  }
+  def bug(): Unit = {
+    val o = new DynamicFields
+    o.foo = "bar"
+  }
+}

